src/python.cpp

```diff
@@ -1980,8 +1980,23 @@ PyObjectRef py_dict_impl(const List& keys, const List& items, bool convert) {
 
 // [[Rcpp::export]]
 SEXP py_dict_get_item(PyObjectRef dict, RObject key) {
-  PyObjectPtr pyKey(r_to_py(key, dict.convert()));
+  
+  PyObject* pyKey(r_to_py(key, dict.convert()));
   PyObject* item = PyDict_GetItem(dict, pyKey);
```

Comment:
Shouldn't we perform `PyDict_Check(dict)` before we attempt treating the `dict` object as a builtin dictionary?

Reply:
It also looks like we might need `PyDict_CheckExact()` here. See:

https://github.com/python/cpython/blob/da6ce58dd5ac109485af45878fca6bfd265b43e9/Python/ceval.c#L2294-L2297

for example code in Python's sources.

Reply:
The weird thing is, `PyDict_Check` is false for `OrderedDict`s ... (which it shouldn't be, right, unless `OrderedDict` is not a proper subclass?))

First, I had the logic the other way round, directly starting with `PyDict_Check` and the case where the check would return false ... but then, the `OrderedDict` test would fail (perhaps because of problems with `__getitem__` and keys that aren't strings?)

Anyway, so I reversed the logic and ended up with that slightly strange way of checking for `NULL` etc (which seemed error-prone to me though)